motr/client.c

```diff
@@ -830,84 +831,112 @@ M0_INTERNAL int m0_op_init(struct m0_op *op,
 
-	/* This call is for first data unit, need to initialize prev_digest*/
-	if (flag & M0_PI_CALC_UNIT_ZERO) {
-		MD5_Init(&pi->prev_digest);
-	}
 
-	/* memcpy, so that we do not change the prev_digest */
-	memcpy(curr_digest, &pi->prev_digest, sizeof(MD5_CTX));
-
-	/* get the curr digest i by updating it*/
+	/* get the curr context i by updating it*/
 	for (i = 0; i < bvec->ov_vec.v_nr; i++) {
-		MD5_Update(curr_digest, bvec->ov_buf[i], bvec->ov_vec.v_count[i]);
+		rc = MD5_Update((MD5_CTX *)curr_context, bvec->ov_buf[i],
+				bvec->ov_vec.v_count[i]);
+		if (rc != 1) {
+			M0_LOG(M0_ERROR, "MD5_Update failed with error %d", rc);
+			return M0_ERR(rc);
+		}
 	}
 
 	/* if caller want checksum without seed and with seed, in this case
 	 * 'pi_value_without_seed' will be used to send wihout seed checksum.
 	 */
-	if (pi_value_without_seed && (flag & M0_PI_CALC_FINAL)) {
-		unsigned char digest[sizeof(MD5_CTX)];
-		memcpy(&digest, curr_digest, sizeof(MD5_CTX));
+	if (pi_value_without_seed) {
+		MD5_CTX context;
+		memcpy((void *)&context, (void *)curr_context, sizeof(MD5_CTX));
 
 		/* 
-		 * NOTE: MD5_final() changes the digest itself and curr_digest
+		 * NOTE: MD5_final() changes the context itself and curr_context
 		 * should not be finalised, thus copy it and use it for MD5_final
 		 */
 
-		MD5_Final(pi_value_without_seed, digest);
+		rc = MD5_Final(pi_value_without_seed, &context);
+		if (rc != 1) {
+			M0_LOG(M0_ERROR, "MD5_Final failed with error %d", rc);
+			return M0_ERR(rc);
+		}
 	}
 
-	/* if seed is passed, memcpy and update the digest calculated so far.
+	/* if seed is passed, memcpy and update the context calculated so far.
 	 * calculate checksum with seed, set the pi_value with seeded checksum.
-	 * NOTE: curr_digest will always have digest without seed.
+	 * If seed is not passed than memcpy context and calculate checksum
+	 * without seed, set the pi_value with unseeded checksum.
+	 * NOTE: curr_context will always have context without seed.
 	 */
+	MD5_CTX context;
+	memcpy((void *)&context, (void *)curr_context, sizeof(MD5_CTX));
+
 	if (seed) {
-		unsigned char seed_digest[sizeof(MD5_CTX)];
-		memcpy(&seed_digest, curr_digest, sizeof(MD5_CTX));
 
-		MD5_Update(&seed_digest, &seed, sizeof(struct m0_pi_seed));
-		MD5_Final(pi->pi_value, &seed_digest);
+		/*
+		 * seed_str will have string represention for 3 uint64_t(8 bytes)
+		 * range for uint64_t is 0 to 18,446,744,073,709,551,615 at max 20
+		 * chars per var, for three var it will be 3*20, +1 for null char.
+		 * seed_str needs to be 61 bytes, round off and taking 64 bytes.
+		 */
+		char seed_str[64] = {'\0'};
+		snprintf(seed_str, sizeof(seed_str), "%"PRIx64"%"PRIx64"%"PRIx64,
+				seed->obj_id.f_container, seed->obj_id.f_key,
+				seed->data_unit_offset);
+		M0_LOG(M0_DEBUG, "f_container %"PRIx64"f_key %"PRIx64
+				"data_unit_offset %"PRIx64"seed_str %s",
+				seed->obj_id.f_container, seed->obj_id.f_key,
+				seed->data_unit_offset, (char *)seed_str);
```

Comment:
The content is already in "seed_str".
This same will be duplicated in the final log?

Reply:
From the string itself its difficult to figure out container/key/offset. Thus I added all of these and string.
Debugger can check if string is created properly and if all args were correct.